### PR TITLE
pry-bond is now compatible to pry/pry@193a229

### PR DESCRIPTION
**Pull request by kaibioinfo**
- Pry::BondCompleter is now a class that is initialized with a Pry instance (to deal with pry/pry@193a229)
- Bond does not use Readline directly, anymore. Instead, Bond gets its input from Pry#complete
- But Bond uses still a global state. The local Pry instance is set before each completion call (using a Mutex to deal with parallel completion requests)

### CODE REVIEW

File: lib/pry/bond/completer.rb [hunk 1]
```diff
@@ -1,21 +1,66 @@
...
+        begin
+          @current_pry_instance = instance
+          yield
+        ensure
+          @current_pry_instance = nil 
+        end
+      end
+    end
+
+    def initialize pry_instance
```
Comment (ghost) [inline lib/pry/bond/completer.rb]: the signature of this method should be the same as, https://github.com/pry/pry/blob/master/lib/pry/input_completer.rb#L44. the first argument should be an input, like `Readline`. i would like to keep the same instance variable name(`@pry`) as well.

File: lib/pry/bond/completer.rb [hunk 2]
```diff
@@ -1,21 +1,66 @@
...
+      BondCompleter.setup if BondCompleter.bond.nil?
+      @pry_instance = pry_instance
+    end
+
+    def call(input, options)
```
Comment (ghost) [inline lib/pry/bond/completer.rb]: we should call `input` something like `str` here, it's too easy to confuse with `_pry_.input` and that concept in general, it is called `str` on pry: https://github.com/pry/pry/blob/master/lib/pry/input_completer.rb#L54

File: lib/pry/bond/completer.rb [hunk 3]
```diff
@@ -1,21 +1,66 @@
...
+        # do nothing as we let pry do the completion
+      end
+
+      def self.line_buffer
+        "" # we give the current line directly to bond
+      end
+    end
+
+    def self.setup(config={})
+      @lock = Mutex.new
```
Comment (ghost) [inline lib/pry/bond/completer.rb:19]: i'm generally not a fan of locks and storing `#bond` on a class is very troublesome.

**Comment by ghost** [comment]
this looks good. i'll check it out locally and i might have some changes of my own as well.

**Comment by ghost** [comment]
hey @kaibioinfo you seem to have an interest in pry&bond, you should have read/write access to this repo now.

**Comment by ghost** [comment]
hey @kaibioinfo - we dont want to use a lock here.

**Comment by kaibioinfo** [comment]
Hi,

I think it is not possible to use Bond in a thread-safe manner. Bond 
stores the eval binding in a class instance variable. So if two pry 
sessions concurrently start a completion they have to be synchronized. I 
have forked the Bond library and will try to make it thread-safety. But 
I don't see another solution currently.

**Comment by ghost** [comment]
could we try to possibly fix bond then? as far as i know, it is not actively maintained a lot anymore. the problem with a lock is that it makes sure when parallelism happens it ends up becoming non-parallel because one thread is waiting on the other. solving it without locks, if possible, would be great.

pry has took the approach of using state through the program as local variables in an effort to avoid global state as much as possible and therefore hopefully be much easier to be thread safe.

**Comment by ghost** [comment]
that's pry master by the way, pry the gem would be atrocious if threads were around.

**Pull request merged**